text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,801 @@
+
+```rust
+trait Foo<T> {
+  fn foo() -> T;
+}
+
+struct Bar<T>(T);
+
+impl<T> Foo<T> for Bar<T> {
+  default fn foo() -> T {
+    static MSG: &str = "sry, no can do";
+    panic!("{}", MSG)
+  }
+}
+
+impl<T: Default> Foo<T> for Bar<T> {
+  fn foo() -> T {
+    static MSG: &str = "it's a go!";
+    println!("{}", MSG);
+    T::default()
+  }
+}
+
+```
+
+Notice that, if one just omits the type arguments, both `MSG` statics have the path `<Bar as Foo>::foo::MSG`. However, we cannot disambiguate by adding type arguments, because we don't have any concrete types to substitute the arguments for. Therefore, we have to disambiguate the `impls`. Since trait specialization is an unstable feature of Rust and the details are in flux, this RFC does not try to provide a mangling based on the `where` clauses of the specialized `impls`. Instead it proposes to re-use the "numeric disambiguator" form already used for macro hygiene and closures. Thus, conflicting `impls` would be disambiguated via an implementation defined suffix, as in `<Bar as Foo>'1::foo::MSG` and `<Bar as Foo>'2::foo::MSG`. This encoding introduces minimal additional syntax and can be replaced with something more human-readable once the definition of trait specialization is final.
+
+
+### Unicode Identifiers
+
+Rust allows Unicode identifiers but our character set is restricted to ASCII alphanumerics, and `_`. In order to transcode the former to the latter, we use the same approach as Swift, which is: encode all non-ascii identifiers via [Punycode][punycode], a standardized and efficient encoding that keeps encoded strings in a rather human-readable format. So for example, the string
+
+```
+"Gödel, Escher, Bach"
+```
+
+is encoded as
+
+```
+"Gdel, Escher, Bach-d3b"
+```
+
+which, as opposed to something like _Base64_, still gives a pretty good idea of what the original string looked like.
+
+Each component of a name, i.e. anything that starts with the number of bytes to read in the examples above, is encoded individually. Components encoded this way are augmented with a `u` suffix so that demanglers know that the identifier needs further decoding. As an example, the function:
+
+```rust
+mod gödel {
+  mod escher {
+    fn bach() {}
+  }
+}
+```
+
+would be mangled as:
+
+```
+_RN15mycrate_4a3b56d8gdel_Fqau6escher4bachVE`
+                    <-------->
+                Unicode component
+```
+
+### Compression/Substitution
+
+The length of symbol names has an influence on how much work compiler, linker, and loader have to perform. The shorter the names, the better. At the same time, Rust's generics can lead to rather long names (which are often not visible in the code because of type inference and `impl Trait`). For example, the return type of the following function:
+
+```rust
+fn quux(s: Vec<u32>) -> impl Iterator<Item=(u32, usize)> {
```

Comment:
```suggestion
fn quux(s: Vec<u32>) -> impl Iterator<Item = (u32, usize)> {
```

Reply:
This is an interesting point, in trait objects, the compiler (and demangler I wrote) still print without spaces around the `=`, e.g. `dyn Iterator<Item=String>`.
Should we change it?

Reply:
>  Should we change it?

Yes. Try rustfmt on: `fn foo() -> impl Iterator<Item=u8> { 0..10 }`

Reply:
I've applied this change to the PR.